pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/BlockingPullSubscriber.java

```diff
@@ -0,0 +1,128 @@
+import com.google.api.core.ApiService.State;
+import com.google.cloud.pubsublite.SequencedMessage;
+import com.google.cloud.pubsublite.cloudpubsub.FlowControlSettings;
+import com.google.cloud.pubsublite.internal.CheckedApiException;
+import com.google.cloud.pubsublite.internal.ExtractStatus;
+import com.google.cloud.pubsublite.internal.wire.Subscriber;
+import com.google.cloud.pubsublite.internal.wire.SubscriberFactory;
+import com.google.cloud.pubsublite.proto.FlowControlRequest;
+import com.google.cloud.pubsublite.proto.SeekRequest;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class BlockingPullSubscriber implements AutoCloseable {
+
+  private final Subscriber underlying;
+
+  private final Lock lock = new ReentrantLock();
+  private final Condition notEmpty = lock.newCondition();
+
+  @GuardedBy("lock")
+  private Optional<CheckedApiException> error = Optional.empty();
+
+  @GuardedBy("lock")
+  private Deque<SequencedMessage> messages = new ArrayDeque<>();
+
+  public BlockingPullSubscriber(
+      SubscriberFactory factory, FlowControlSettings settings, SeekRequest initialSeek)
+      throws CheckedApiException {
+    underlying = factory.newSubscriber(this::addMessages);
+    underlying.addListener(
+        new Listener() {
+          @Override
+          public void failed(State state, Throwable throwable) {
+            fail(ExtractStatus.toCanonical(throwable));
+          }
+        },
+        MoreExecutors.directExecutor());
+    underlying.startAsync().awaitRunning();
+    try {
+      underlying.seek(initialSeek).get();
+    } catch (InterruptedException e) {
+      throw ExtractStatus.toCanonical(e);
+    } catch (ExecutionException e) {
+      throw ExtractStatus.toCanonical(e.getCause());
+    }
+    underlying.allowFlow(
+        FlowControlRequest.newBuilder()
+            .setAllowedMessages(settings.messagesOutstanding())
+            .setAllowedBytes(settings.bytesOutstanding())
+            .build());
+  }
+
+  private void fail(CheckedApiException e) {
+    lock.lock();
+    try {
+      error = Optional.of(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void addMessages(Collection<SequencedMessage> new_messages) {
+    lock.lock();
+    try {
+      messages.addAll(new_messages);
+      if (!messages.isEmpty()) {
+        notEmpty.signal();
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  // Blocking call until there is at least one message to return.
+  public SequencedMessage pull() throws InterruptedException, CheckedApiException {
```

Comment:
can you make this signature:

// Returns an empty optional on timeout.
public Optional<SequencedMessage> pull(Duration maxWait) throws CheckedApiException;

Then it can be easily reused to pull batches by calling pull(Duration.ZERO) in a loop.

Reply:
You also probably want to ensure there's only one call to "pull" outstanding at once: Can you check and set a boolean when you enter a "pull" call and clear it when you leave?

Reply:
Added comments in the interface.

Reply:
Please make this change.

Reply:
Discussed offline, we will make interface to be onData and messageifavailable.